File: agricultural-crop-management-backend/src/main/java/org/example/QuanLyMuaVu/Entity/Task.java
package org.example.QuanLyMuaVu.Entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.FieldDefaults;
import org.example.QuanLyMuaVu.Enums.TaskStatus;

import java.time.LocalDate;
import java.time.LocalDateTime;

@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
@FieldDefaults(level = AccessLevel.PRIVATE)
@Entity
@Table(name = "tasks")
public class Task {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "task_id")
    Integer id;

    @ManyToOne
    @JoinColumn(name = "user_id", nullable = false)
    User user;

    /**
     * Optional link to a season. New season operations APIs will always set this;
     * legacy APIs may leave it null.
     */
    @ManyToOne
    @JoinColumn(name = "season_id")
    Season season;

    @Column(nullable = false)
    String title;

    @Column(columnDefinition = "TEXT")
    String description;

    /**
     * Planned date of the task within the season.
     */
    @Column(name = "planned_date")
    LocalDate plannedDate;

    @Column(name = "due_date")
    LocalDate dueDate;

    @Enumerated(EnumType.STRING)
    @Column(name = "status")
    TaskStatus status;

    @Column(name = "actual_start_date")
    LocalDate actualStartDate;

    @Column(name = "actual_end_date")
    LocalDate actualEndDate;

    @Column(name = "notes", columnDefinition = "TEXT")
    String notes;

    @Column(name = "created_at", columnDefinition = "TIMESTAMP DEFAULT CURRENT_TIMESTAMP")
    LocalDateTime createdAt;
}

File: ACM Web Platform Design System/src/types/Task.ts
export type TaskStatus = 'PENDING' | 'IN_PROGRESS' | 'DONE' | 'OVERDUE' | 'CANCELLED';

export interface Task {
  id: number;
  title: string;
  description?: string;
  seasonId?: number;
  seasonName?: string;
  plannedDate?: string;
  dueDate?: string;
  actualStartDate?: string;
  actualEndDate?: string;
  status: TaskStatus;
  notes?: string;
  isOverdue?: boolean;
  createdAt?: string;
}

export interface CreateTaskRequest {
  seasonId?: number;
  title: string;
  description?: string;
  plannedDate?: string;
  dueDate?: string;
  notes?: string;
}

export interface UpdateTaskRequest {
  title?: string;
  description?: string;
  plannedDate?: string;
  dueDate?: string;
  notes?: string;
}

export interface CompleteTaskRequest {
  actualEndDate: string;
  completionNotes?: string;
}

export interface StartTaskRequest {
  actualStartDate?: string;
}

export interface TaskSearchParams {
  seasonId?: number;
  status?: TaskStatus;
  from?: string;
  to?: string;
  q?: string;
  page?: number;
  size?: number;
}
